src/tactic/equiv_rw.lean

```diff
@@ -0,0 +1,260 @@
+  -- The function arrow is technically a bifunctor `Typeᵒᵖ → Type → Type`,
+  -- but the pattern matcher will never see this.
+  `equiv.arrow_congr',
+  -- Allow rewriting in subtypes:
+  `equiv.subtype_equiv_of_subtype',
+  -- Allow rewriting in the first component of a sigma-type:
+  `equiv.sigma_congr_left',
+  -- Allow rewriting ∀s:
+  -- (You might think that repeated application of `equiv.forall_congr'
+  -- would handle the higher arity cases, but unfortunately unification is not clever enough.)
+  `equiv.forall₃_congr',
+  `equiv.forall₂_congr',
+  `equiv.forall_congr',
+  -- Allow rewriting in argument of Pi types:
+   `equiv.Pi_congr_left',
+  -- Handles `sum` and `prod`, and many others:
+  `bifunctor.map_equiv,
+  -- Handles `list`, `option`, and many others:
+  `functor.map_equiv,
+  -- We have to filter results to ensure we don't cheat and use exclusively `equiv.refl` and `iff.refl`!
+  `equiv.refl,
+  `iff.refl
+  ].mmap (λ n, try_core (mk_const n)),
+  return (exprs.map option.to_list).join -- TODO: implement `.mfilter_map mk_const`?
+
+declare_trace adapt_equiv
+
+/-- Implementation of `adapt_equiv`, using `solve_by_elim`. -/
+meta def adapt_equiv_core (eq ty : expr) : tactic expr :=
+do
+  -- We prepare a synthetic goal of type `(%%ty ≃ _)`, for some placeholder right hand side.
+  initial_goals ← get_goals,
+  g ← to_expr ``(%%ty ≃ _) >>= mk_meta_var,
+  set_goals [g],
+  -- Assemble the relevant lemmas.
+  equiv_congr_lemmas ← equiv_congr_lemmas,
+  /-
+    We now call `solve_by_elim` to try to generate the requested equivalence.
+    There are a few subtleties!
+    * We make sure that `eq` is the first lemma, so it is applied whenever possible.
+    * In `equiv_congr_lemmas`, we put `equiv.refl` last so it is only used when it is not possible
+      to descend further.
+    * To avoid the possibility that the entire resulting expression is built out of
+      congruence lemmas and `equiv.refl`, we use the `accept` subtactic of `solve_by_elim`
+      to reject any results which neither contain `eq` or a remaining metavariable.
+    * Since some congruence lemmas generate subgoals with `∀` statements,
+      we use the `pre_apply` subtactic of `solve_by_elim` to preprocess each new goal with `intros`.
+  -/
+  solve_by_elim {
+    use_symmetry := false,
+    use_exfalso := false,
+    lemmas := some (eq :: equiv_congr_lemmas),
+    -- TODO decide an appropriate upper bound on search depth.
+    max_steps := 6,
+    -- Subgoals may contain function types,
+    -- and we want to continue trying to construct equivalences after the binders.
+    pre_apply := tactic.intros >> skip,
+    discharger := trace_if_enabled `adapt_equiv "Failed, no congruence lemma applied!" >> failed,
+    -- We accept any branch of the `solve_by_elim` search tree which
+    -- either still contains metavariables, or already contains at least one copy of `eq`.
+    -- This is to prevent generating equivalences built entirely out of `equiv.refl`.
```

Comment:
I'm not sure I understand this condition.

I'm a bit confused by counting the number of `eq`s.  In general we might want multiple copies of `eq` in the equivalence, right?  For example if we rewrite `α ≃ β` in `list (α × α)`.

Is the purpose of this handler to make sure that a subgoal `ℕ × ℕ ≃ _` is solved using `equiv.refl` instead of `bifunctor.map_equiv equiv.refl equiv.refl`?  Wouldn't it be more appropriate to apply `equiv.refl` in `pre_apply` if the subgoal doesn't contain `α`?

It's also a bit confusing that the `accept` functions handles multiple original goals.  There's always just one, right?

Reply:
1. The purpose of this handler is to ensure that we _don't_ just use `equiv.refl` on everything, and instead require that `eq` appears _at least once_ in the final equivalence.

2. But your example indicates a real problem in my algorithm, that I'll have to think about. I want to avoid using `equiv.refl` when `eq` could be used on a sub-term, so for this reason I've put ` but I'd really prefer that we don't unnecessarily generate monstrosities such as `bifunctor.map_equiv equiv.refl equiv.refl`. At the moment, in my understanding, that's actually what we do! Maybe that's okay, maybe I can work out how to avoid doing that (although I don't see it yet, but perhaps `accept` can do this too), and maybe some form of simplification after the fact would be appropriate.

3. It is a bit confusing that `accept` handles multiple original goals, but since other applications of `solve_by_elim` (particularly `library_search`) relying on the `backtrack_all_goals` setting, I think `accept` has to deal with this scenario too, even though it is irrelevant in the application in this branch.

Reply:
Regarding point 2. I've added a step that simplifies the generated equivalence before returning it, and this successfully compresses away unnecessary `map_equiv equiv.refl` type subexpressions to `equiv.refl`.

Reply:
> The purpose of this handler is to ensure that we don't just use equiv.refl on everything, and instead require that eq appears at least once in the final equivalence.

Isn't it enough to put `equiv.refl` at the end of `equiv_congr_lemmas`?  I would have assumed that `solve_by_elim` would try the other lemmas first (and e.g. for `_ × _` applying the congruence lemma for product would always succeed).

Reply:
You're absolutely right. It's enough to replace using `accept` with a simple test that checks that we used the given equivalence. I've done this.